### PR TITLE
Remove IActiveStatementSpanTracker

### PR DESCRIPTION
**Pull request by sharwell**
Instead of using a workspace service to retrieve the spans the EnC analyzer now takes a snapshot of the spans along with the corresponding document snapshot. Methods on EditAndContinueWorkspaceService are updated to take a callback that retrieves the spans for a given document on demand. This callback is called just before the service queries the EnC analyzer for an analysis of a document but at the same time during the EditAndContinueWorkspaceService method call that passed in the callback. This is important because in future PR each method of the service will be a remote call to OOP and the solution snapshot this method operates on is only going to be available to the OOP within the method call.

### CODE REVIEW

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 1]
```diff
@@ -440,7 +440,10 @@ private AsyncLazy<DocumentAnalysisResults> GetDocumentAnalysisNoLock(Document? b
                             documentBaseActiveStatements = ImmutableArray<ActiveStatement>.Empty;
                         }
 
-                        return await analyzer.AnalyzeDocumentAsync(baseDocument, documentBaseActiveStatements, document, cancellationToken).ConfigureAwait(false);
+                        var oldSpansInNewDocumentBuilder = ImmutableArray.CreateBuilder<TextSpan?>(documentBaseActiveStatements.Length);
+                        oldSpansInNewDocumentBuilder.Count = documentBaseActiveStatements.Length;
+                        var oldSpansInNewDocument = oldSpansInNewDocumentBuilder.MoveToImmutable();
+                        return await analyzer.AnalyzeDocumentAsync(baseDocument, documentBaseActiveStatements, document, oldSpansInNewDocument, cancellationToken).ConfigureAwait(false);
```
Comment (sharwell) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs:446]: @tmat This portion is the part I didn't understand for this change

File: src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs [hunk 2]
```diff
@@ -2323,11 +2316,11 @@ public async Task ActiveStatements()
...
-            var spanTracker = Assert.IsType<TestActiveStatementSpanTracker>(workspace.Services.GetRequiredService<IActiveStatementSpanTrackerFactory>().GetOrCreateActiveStatementSpanTracker());
-            spanTracker.Spans = new Dictionary<DocumentId, TextSpan?[]>
-            {
-                { documentId, new TextSpan?[] { activeSpan11, activeSpan12 } }
-            };
+            //var spanTracker = Assert.IsType<TestActiveStatementSpanTracker>(workspace.Services.GetRequiredService<IActiveStatementSpanTrackerFactory>().GetOrCreateActiveStatementSpanTracker());
+            //spanTracker.Spans = new Dictionary<DocumentId, TextSpan?[]>
+            //{
+            //    { documentId, new TextSpan?[] { activeSpan11, activeSpan12 } }
+            //};
```
Comment (sharwell) [inline src/EditorFeatures/Test/EditAndContinue/EditAndContinueWorkspaceServiceTests.cs:2323]: 📝 Commented code in this file needs to be removed

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 3]
```diff
@@ -21,8 +23,6 @@
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
-#nullable enable
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:24]: Why remove nullable? Do we enable it now for the whole project?

Comment (sharwell) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:24]: I didn't remove it. I fixed the misplaced directive.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -373,6 +370,7 @@ protected virtual string GetSuspensionPointDisplayName(SyntaxNode node, EditKind
             Document? oldDocument,
             ImmutableArray<ActiveStatement> baseActiveStatements,
             Document document,
+            ImmutableArray<TextSpan?> oldSpansInNewDocument,
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:373]: >oldSpansInNewDocument [](start = 38, length = 21)

I'd call this `documentActiveStatementSpans` in all methods

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 5]
```diff
@@ -645,8 +641,8 @@ private static bool ReportFatalErrorAnalyzeDocumentAsync(ImmutableArray<ActiveSt
             Dictionary<SyntaxNode, EditKind> editMap,
             SourceText oldText,
             SourceText newText,
-            DocumentId documentId,
             ImmutableArray<ActiveStatement> oldActiveStatements,
+            ImmutableArray<TextSpan?> oldSpansInNewText,
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:645]: >oldSpansInNewText [](start = 38, length = 17)

Call this `newTextActiveStatements`

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 6]
```diff
@@ -919,13 +917,14 @@ public LambdaInfo WithMatch(Match<SyntaxNode> match, SyntaxNode newLambdaBody)
...
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:921]: >oldSpansInNewText [](start = 38, length = 17)

`newTextActiveStatements`

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 7]
```diff
@@ -674,11 +670,12 @@ private static bool ReportFatalErrorAnalyzeDocumentAsync(ImmutableArray<ActiveSt
             Match<SyntaxNode> topMatch,
             SourceText oldText,
             SourceText newText,
-            DocumentId documentId,
             ImmutableArray<ActiveStatement> oldActiveStatements,
+            ImmutableArray<TextSpan?> oldSpansInNewText,
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:674]: >oldSpansInNewText [](start = 38, length = 17)

`newActiveStatementSpans`

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 3 above)
Comment (tmat) [reply src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:24]: I see.

**Pull request merged**